Software/Drivers/Src/Contactors.c

```diff
@@ -0,0 +1,134 @@
+#include "Contactors.h"
+
+static SemaphoreHandle_t contactorsMutex = NULL;
+static StaticSemaphore_t contactorsMutexBuffer;
+
+// array to hold the contactor structs
+static contactor_t contactors[NUM_CONTACTORS];
+
+void init_contactors() {
+    GPIO_InitTypeDef GPIO_InitStruct = { 0 };
+
+    // Enable clock for GPIO A/B/C 
+    __HAL_RCC_GPIOA_CLK_ENABLE();
+    __HAL_RCC_GPIOB_CLK_ENABLE();
+    __HAL_RCC_GPIOC_CLK_ENABLE();
+
+    // init channel A contactor sense pins 
+    GPIO_InitStruct.Pin = HV_POSITIVE_SENSE_PIN_NUM | HV_NEGATIVE_SENSE_PIN_NUM | ARRAY_SENSE_PIN_NUM;
+    GPIO_InitStruct.Mode = GPIO_MODE_INPUT;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+    HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
+
+    // init channel B contactor sense pins
+    GPIO_InitStruct.Pin = ARRAY_PRE_SENSE_PIN_NUM;
+    GPIO_InitStruct.Mode = GPIO_MODE_INPUT;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+    HAL_GPIO_Init(GPIOB, &GPIO_InitStruct);
+
+    // init channel B contactor control pins
+    GPIO_InitStruct.Pin = HV_NEGATIVE_CONTROL_PIN_NUM | ARRAY_CONTROL_PIN_NUM | ARRAY_PRE_CONTROL_PIN_NUM;
+    GPIO_InitStruct.Mode = GPIO_MODE_OP_PP;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+    HAL_GPIO_Init(GPIOB, &GPIO_InitStruct);
+
+    // init channel C contactor  control pins
+    GPIO_InitStruct.Pin = HV_POSITIVE_CONTROL_PIN_NUM; 
+    GPIO_InitStruct.Mode = GPIO_MODE_OP_PP;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+    HAL_GPIO_Init(GPIOC, &GPIO_InitStruct);
+
+    // initializing the pindef into contactor structs
+    contactors[HV_Pos_contactor] = (contactor_t){   
+        .control_pin = { HV_POSITIVE_CONTROL_PORT, HV_POSITIVE_CONTROL_PIN_NUM },
```

Comment:
good use of explicit struct member definitions with the `.variable = value`, keep it up